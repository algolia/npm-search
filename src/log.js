import bunyan from 'bunyan';
import bunyanDebugStream from 'bunyan-debug-stream';

const stream = bunyanDebugStream({
  basepath: __dirname,
});

const logger = bunyan.createLogger({
  name: 'npm-search',
  streams: [
    {
      level: 'debug',
      type: 'raw',
      stream,
    },
  ],
  serializers: bunyanDebugStream.serializers,
});

export default logger;
