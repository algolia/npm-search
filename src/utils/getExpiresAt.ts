import { config } from '../config';

export function getExpiresAt(popular = false): number {
  // Rand -48h to +48h, to spread refresh
  const randHours = Math.floor(Math.random() * (-48 - 48 + 1)) + 48;

  // Round minutes to avoid too many values in facet
  const minutes = Math.ceil(Math.floor(Math.random() * (60 + 1)) / 30) * 30;

  const expiresAt = new Date(
    Date.now() + (popular ? config.popularExpiresAt : config.expiresAt)
  );
  expiresAt.setMilliseconds(0);
  expiresAt.setSeconds(0);
  expiresAt.setMinutes(minutes);
  expiresAt.setHours(randHours);

  return expiresAt.getTime();
}
