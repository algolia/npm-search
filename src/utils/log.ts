import bunyan from 'bunyan';
import bunyanDebugStream from 'bunyan-debug-stream';

const stream = bunyanDebugStream({
  showDate: process.env.NODE_ENV !== 'production',
  showProcess: false,
  showLoggerName: false,
  showPid: process.env.NODE_ENV !== 'production',
});

export const log = bunyan.createLogger({
  name: 'npm-search',
  streams: [
    {
      level: 'info',
      type: 'raw',
      stream,
    },
  ],
  serializers: bunyanDebugStream.serializers,
});
