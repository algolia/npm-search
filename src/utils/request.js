import http from 'http';
import https from 'https';

import got from 'got';

const options = {
  keepAlive: true,
  timeout: 60000,
  maxFreeSockets: 2000,
  scheduling: 'fifo',
};

// The agents will pool TCP connections
export const httpAgent = new http.Agent(options);
export const httpsAgent = new https.Agent(options);

export async function request(url, opts) {
  return await got(url, {
    ...opts,
    agent: {
      http: httpAgent,
      https: httpsAgent,
    },
  });
}
