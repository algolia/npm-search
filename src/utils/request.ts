import http from 'http';
import https from 'https';

import type { OptionsOfJSONResponseBody } from 'got';
import got from 'got';

// eslint-disable-next-line @typescript-eslint/no-var-requires, import/no-commonjs
const { version } = require('../../package.json');

export const USER_AGENT = `Algolia npm-search/${version} (https://github.com/algolia/npm-search)`;

const options = {
  keepAlive: true,
  timeout: 60000,
  maxFreeSockets: 2000,
  scheduling: 'fifo',
};

// The agents will pool TCP connections
export const httpAgent = new http.Agent(options);
export const httpsAgent = new https.Agent(options);

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export async function request<TRes>(
  url: string,
  opts: OptionsOfJSONResponseBody
) {
  return await got<TRes>(url, {
    ...opts,
    headers: {
      ...(opts.headers || {}),
      'user-agent': USER_AGENT,
    },
    timeout: 15000,
    dnsCache: true,
    dnsLookupIpVersion: 'ipv4',
    agent: {
      http: httpAgent,
      https: httpsAgent,
    },
  });
}
