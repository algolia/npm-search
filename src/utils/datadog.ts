import StatsD from 'hot-shots';

import { log } from './log';

const env = process.env.NODE_ENV === 'production' ? 'prod' : 'dev';

export const datadog = new StatsD({
  host: process.env.DOGSTATSD_HOST || 'localhost',
  port: 8125,
  prefix: 'alg.npmsearch.',
  globalTags: {
    env,
  },
  errorHandler(error): void {
    log.error('[DATADOG ERROR]', error);
  },
});
