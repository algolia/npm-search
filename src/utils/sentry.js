import Sentry from '@sentry/node';

import log from '../log.js';

Sentry.init({
  dsn: process.env.SENTRY_DSN,
  release: `1.0.0`,
  environment: 'prod',
  serverName: 'npm-search',
});

export function report(err, extra = {}) {
  log.error(err.message);
  if (!process.env.SENTRY_DSN) {
    log.error(err);
    return;
  }

  Sentry.withScope((scope) => {
    scope.setExtras(extra);
    Sentry.captureException(err);
  });
}

export function drain() {
  const client = Sentry.getCurrentHub().getClient();
  if (client) {
    return client.close(2000);
  }
  return Promise.resolve();
}
