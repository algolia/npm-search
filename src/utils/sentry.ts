import * as Sentry from '@sentry/node';

import { log } from './log';

Sentry.init({
  dsn: process.env.SENTRY_DSN,
  release: `1.0.0`,
  environment: 'prod',
  serverName: 'npm-search',
  ignoreErrors: [
    /error happened in your connection/,
    /503 Service Unavailable/,
    /<!DOCTYPE html>/,
  ],
});

export function report(err, extra = {}): void {
  log.error(err.message);
  if (!process.env.SENTRY_DSN) {
    log.error(err);
    return;
  }

  Sentry.withScope((scope) => {
    scope.setExtras(extra);
    Sentry.captureException(err);
  });
}

export async function drain(): Promise<boolean> {
  const client = Sentry.getCurrentHub().getClient();
  if (client) {
    return await client.close(2000);
  }
  return true;
}
