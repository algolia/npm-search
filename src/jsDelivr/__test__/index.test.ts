import { log } from '../../utils/log';
import * as api from '../index';

jest.mock('../../utils/log', () => {
  return {
    log: {
      info: jest.fn(),
      warn: jest.fn(),
      error: jest.fn(),
    },
  };
});

jest.setTimeout(10000);

// eslint-disable-next-line jest/require-top-level-describe
beforeEach(() => {
  jest.resetAllMocks();
});

describe('hits', () => {
  describe('getHits()', () => {
    beforeAll(() => {
      api.hits.clear();
      api.hits.set('jquery', 1234);
    });

    it('should get one formatted hit', () => {
      expect(api.getHits([{ name: 'jquery' }])).toEqual([
        {
          jsDelivrHits: 1234,
          _searchInternal: {
            jsDelivrPopularity: 1,
          },
        },
      ]);
    });
    it('should get multiple formatted hits', () => {
      expect(
        api.getHits([{ name: 'jquery' }, { name: 'thispackagedoesnotexist' }])
      ).toEqual([
        {
          jsDelivrHits: 1234,
          _searchInternal: {
            jsDelivrPopularity: 1,
          },
        },
        {
          jsDelivrHits: 0,
          _searchInternal: {
            jsDelivrPopularity: 0,
          },
        },
      ]);
    });
  });

  describe('loadHits()', () => {
    beforeAll(async () => {
      await api.loadHits();
    });
    it('should download all packages hits', () => {
      expect(api.hits.size).toBeGreaterThan(60000); // 66790 (2019-08)
    });

    it('should get one hit', () => {
      expect(api.hits.get('jquery')).toBeGreaterThan(1000000000); // 1065750968 (2019-08)
    });

    it('should not get one hit', () => {
      expect(api.hits.get('thispackagedoesnotexist')).toBeUndefined();
    });
  });
});

describe('files', () => {
  describe('getFilesList()', () => {
    it('should get a flat list of files', async () => {
      const files = await api.getFilesList({
        name: 'jest',
        version: '24.8.0',
      });
      expect(files).toMatchSnapshot();
    });

    it('should not get a files list for fake package', async () => {
      const files = await api.getFilesList({
        name: 'thispackagedoesnotexist',
        version: '3.33.0',
      });
      expect(files).toEqual([]);
      expect(log.error.mock.calls[0][0]).toEqual(
        'Failed to fetch https://data.jsdelivr.com/v1/package/npm/thispackagedoesnotexist@3.33.0/flat'
      );
    });
  });

  describe('getAllFilesList()', () => {
    it('should get a flat list of files', async () => {
      const files = await api.getAllFilesList([
        { name: 'jest', version: '24.8.0' },
      ]);
      expect(files).toMatchSnapshot();
    });

    it('should get multiple flat list of files', async () => {
      const files = await api.getAllFilesList([
        {
          name: 'jest',
          version: '24.8.0',
        },
        {
          name: 'thispackagedoesnotexist',
          version: '3.33.0',
        },
      ]);
      expect(files).toMatchSnapshot();
      expect(log.error.mock.calls[0][0]).toEqual(
        'Failed to fetch https://data.jsdelivr.com/v1/package/npm/thispackagedoesnotexist@3.33.0/flat'
      );
    });
  });
});
