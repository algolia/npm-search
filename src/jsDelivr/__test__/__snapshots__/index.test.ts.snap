// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`files getAllFilesList() should get a flat list of files 1`] = `
Array [
  Array [
    Object {
      "hash": "+uxmYs/88pYWmLwFS3M54NGjE+hX6sBmwNOpzaW1LGk=",
      "name": "/bin/jest.js",
      "size": 343,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "MvOGr1Lc6r8wEe8GNmscm3Sx/QWEFE4Is1AZ5rQzFr8=",
      "name": "/build/jest.d.ts",
      "size": 291,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "BEQ5sRqArzHCh5sNbwjxHRQunhxkCD1HXcM9EdYAKPc=",
      "name": "/build/jest.d.ts.map",
      "size": 171,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "m5wVGuVr5Pq4z5L2vpeMVA3rbLV4kQ0MCPuo0newsmY=",
      "name": "/build/jest.js",
      "size": 1030,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "m/vOMvpK2FU19W9PYavnEExEToN7HHU1mb/f/ooU3eQ=",
      "name": "/LICENSE",
      "size": 1099,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "9hWvkPsgtCTc1w0lswu1AO+Q+S19Dppeg5bNklG/Khg=",
      "name": "/package.json",
      "size": 925,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "MPu0d2f8or6adBXZZLUNh6vL7Yeg34MmOBiupdclu10=",
      "name": "/README.md",
      "size": 551,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "CycshPBWVvIRZozw+b1pnAvKYC1Q7aPvcT8tS+HPepU=",
      "name": "/tsconfig.json",
      "size": 162,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "kt2uoTK/NmyQe2OUiNwpdwxV4RxgS2gW9rEgOtj+lZU=",
      "name": "/tsconfig.tsbuildinfo",
      "size": 220798,
      "time": "1985-10-26T08:15:00.000Z",
    },
  ],
]
`;

exports[`files getAllFilesList() should get multiple flat list of files 1`] = `
Array [
  Array [
    Object {
      "hash": "+uxmYs/88pYWmLwFS3M54NGjE+hX6sBmwNOpzaW1LGk=",
      "name": "/bin/jest.js",
      "size": 343,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "MvOGr1Lc6r8wEe8GNmscm3Sx/QWEFE4Is1AZ5rQzFr8=",
      "name": "/build/jest.d.ts",
      "size": 291,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "BEQ5sRqArzHCh5sNbwjxHRQunhxkCD1HXcM9EdYAKPc=",
      "name": "/build/jest.d.ts.map",
      "size": 171,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "m5wVGuVr5Pq4z5L2vpeMVA3rbLV4kQ0MCPuo0newsmY=",
      "name": "/build/jest.js",
      "size": 1030,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "m/vOMvpK2FU19W9PYavnEExEToN7HHU1mb/f/ooU3eQ=",
      "name": "/LICENSE",
      "size": 1099,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "9hWvkPsgtCTc1w0lswu1AO+Q+S19Dppeg5bNklG/Khg=",
      "name": "/package.json",
      "size": 925,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "MPu0d2f8or6adBXZZLUNh6vL7Yeg34MmOBiupdclu10=",
      "name": "/README.md",
      "size": 551,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "CycshPBWVvIRZozw+b1pnAvKYC1Q7aPvcT8tS+HPepU=",
      "name": "/tsconfig.json",
      "size": 162,
      "time": "1985-10-26T08:15:00.000Z",
    },
    Object {
      "hash": "kt2uoTK/NmyQe2OUiNwpdwxV4RxgS2gW9rEgOtj+lZU=",
      "name": "/tsconfig.tsbuildinfo",
      "size": 220798,
      "time": "1985-10-26T08:15:00.000Z",
    },
  ],
  Array [],
]
`;

exports[`files getFilesList() should get a flat list of files 1`] = `
Array [
  Object {
    "hash": "+uxmYs/88pYWmLwFS3M54NGjE+hX6sBmwNOpzaW1LGk=",
    "name": "/bin/jest.js",
    "size": 343,
    "time": "1985-10-26T08:15:00.000Z",
  },
  Object {
    "hash": "MvOGr1Lc6r8wEe8GNmscm3Sx/QWEFE4Is1AZ5rQzFr8=",
    "name": "/build/jest.d.ts",
    "size": 291,
    "time": "1985-10-26T08:15:00.000Z",
  },
  Object {
    "hash": "BEQ5sRqArzHCh5sNbwjxHRQunhxkCD1HXcM9EdYAKPc=",
    "name": "/build/jest.d.ts.map",
    "size": 171,
    "time": "1985-10-26T08:15:00.000Z",
  },
  Object {
    "hash": "m5wVGuVr5Pq4z5L2vpeMVA3rbLV4kQ0MCPuo0newsmY=",
    "name": "/build/jest.js",
    "size": 1030,
    "time": "1985-10-26T08:15:00.000Z",
  },
  Object {
    "hash": "m/vOMvpK2FU19W9PYavnEExEToN7HHU1mb/f/ooU3eQ=",
    "name": "/LICENSE",
    "size": 1099,
    "time": "1985-10-26T08:15:00.000Z",
  },
  Object {
    "hash": "9hWvkPsgtCTc1w0lswu1AO+Q+S19Dppeg5bNklG/Khg=",
    "name": "/package.json",
    "size": 925,
    "time": "1985-10-26T08:15:00.000Z",
  },
  Object {
    "hash": "MPu0d2f8or6adBXZZLUNh6vL7Yeg34MmOBiupdclu10=",
    "name": "/README.md",
    "size": 551,
    "time": "1985-10-26T08:15:00.000Z",
  },
  Object {
    "hash": "CycshPBWVvIRZozw+b1pnAvKYC1Q7aPvcT8tS+HPepU=",
    "name": "/tsconfig.json",
    "size": 162,
    "time": "1985-10-26T08:15:00.000Z",
  },
  Object {
    "hash": "kt2uoTK/NmyQe2OUiNwpdwxV4RxgS2gW9rEgOtj+lZU=",
    "name": "/tsconfig.tsbuildinfo",
    "size": 220798,
    "time": "1985-10-26T08:15:00.000Z",
  },
]
`;
